test_util/test_launch.py

```diff
@@ -0,0 +1,310 @@
+    assert 'access_key_id' in info['provider']
+    assert 'secret_access_key' in info['provider']
+    assert info['ssh']['key_name'] == 'foobar'
+    assert info['ssh']['delete_with_stack'] is True
+    assert info['ssh']['private_key'] == MOCK_SSH_KEY_DATA
+    assert info['ssh']['user'] == 'core'
+    # check that description is updated with info
+    assert 'stack_name' in desc
+
+
+def test_no_files_specified(tmpdir, mocked_aws_cf_simple):
+    """Ensure typical usage works without specifying config and info file paths
+    """
+    with tmpdir.as_cwd():
+        config_path = tmpdir.join('config.yaml')
+        config_path.write(mocked_aws_cf_simple)
+        assert main(['create']) == 0
+        assert main(['wait']) == 0
+        assert main(['describe']) == 0
+        assert main(['test']) == 0
+        assert main(['delete']) == 0
+
+
+def test_noop():
+    """Ensure docopt exit (displays usage)
+    """
+    with pytest.raises(DocoptExit):
+        main([])
+    with pytest.raises(DocoptExit):
+        main(['foobar'])
+
+
+def test_conflicting_path(tmpdir, mocked_aws_cf_simple):
+    """Ensure default cluster info path is never overwritten
+    by launching successive clusters
+    """
+    with tmpdir.as_cwd():
+        tmpdir.join('config.yaml').write(mocked_aws_cf_simple)
+        assert main(['create']) == 0
+        assert main(['create']) == 1
+
+
+def test_missing_input(tmpdir):
+    """No files are provided so any operation should fail
+    """
+    with tmpdir.as_cwd():
+        for cmd in ['create', 'wait', 'describe', 'delete', 'test']:
+            with pytest.raises(FileNotFoundError):
+                main([cmd])
+
+
+def mock_stack_not_found(*args):
+    raise Exception('Mock stack was not found!!!')
+
+
+def test_missing_aws_stack(mocked_aws_cf_simple, monkeypatch):
+    """ Tests that clean and appropriate errors will be raised
+    """
+    monkeypatch.setattr(test_util.aws.CfStack, '__init__', mock_stack_not_found)
+    config = yaml.load(mocked_aws_cf_simple)
+    aws_launcher = get_launcher(config['type'], config['provider_info'])
+
+    def check_stack_error(cmd, args):
+        with pytest.raises(LauncherError) as exinfo:
+            getattr(aws_launcher, cmd)(*args)
+        assert exinfo.value.error == 'StackNotFound'
+
+    info = aws_launcher.create(config)
+    check_stack_error('wait', (info,))
+    check_stack_error('describe', (info,))
+    check_stack_error('delete', (info,))
+    check_stack_error('test', (info, 'py.test'))
+
+
+@pytest.mark.usefixtures('mocked_aws_cf_simple')
+def test_aws_ssh_key_handling(mocked_ssh_key_path):
+    provider_info = {
+        'region': 'us-west-2',
+        'access_key_id': 'foo',
+        'secret_access_key': 'bar'}
+    aws_launcher = get_launcher('cloudformation', provider_info)
+    # Test most minimal config (nothing given and key is generated)
+    ssh_info = aws_launcher.ssh_from_config({'stack_name': 'foo'})
+    assert ssh_info['delete_with_stack'] is True
```

Comment:
nit: I find it a lot better to compare the whole dictionary exactly. That way you check for
1. any keys that got added you didn't expect (some testing code that added a key then people downstream start depending on it...)
2. you find out all the keys that are off in a nice diff that py.test gives, rather than the first key which is off only. There's also much nicer errors if ssh_info doesn't contain an expected error (Was on the right side but isn't on the left side and heres the value which was expected) rather than what you'll currently get which is a KeyError exception and none of the keys of the dictoinary printed.